### PR TITLE
add `:doc` command to REPL

### PR DESCRIPTION
**Pull request by paulbutcher**
This implements another of the suggestions within #3989

Draft for now because there are some unanswered questions.

This works reasonably nicely:

Screenshot 2022-07-09 at 11 19 13

But...

I haven't found a way to display the type of a class (so all that's displayed is the documentation):

Screenshot 2022-07-09 at 11 21 27

And it would probably make sense to display its signatures (in the same way as does api.flix.dev).

I *have* managed to find a way to display the signature of an enum, but that relies on using `tpeDeprecated` which probably isn't a good thing?

And the formatting can get messy:

Screenshot 2022-07-09 at 11 24 34

So, to do this properly, we probably need to:

* Do something about formatting markdown nicely for terminal output
* Extract some of the code within the documentor phase to re-use here. Or run the documentor when compiling within the repl and make the results available?

### CODE REVIEW

**Comment by magnus-madsen** [comment]
Looks like a good start! 

We need a collection of helper functions, probably to be placed in `flix/language/fmt` for formatting various things. (https://github.com/flix/flix/issues/2163)

We should not use `tpeDeprecated`. Trying to hard to remove that. There should be a `Scheme sc` that can be used.

**Comment by magnus-madsen** [comment]
We could start with just def/sig/enum support.

**Comment by magnus-madsen** [comment]
Btw. does Clojure have something like this? Does it use colors?

**Comment by paulbutcher** [comment]
> Btw. does Clojure have something like this? Does it use colors?

It does, and no it doesn't use colours:

Screenshot 2022-07-10 at 21 17 44

When I said that we should "Do something about formatting markdown nicely for terminal output", I wasn't thinking of colours, I was more thinking about layout (see the mess that `:doc Applicative` creates above). But it looks like Clojure gets around that by not having long lines in the source documentation (whereas we typically have lines longer than 100 characters).

**Comment by paulbutcher** [comment]
> There should be a Scheme sc that can be used.

I'm not seeing it for an Enum? As far as I can see, Specs are only associated with Defs and Sigs?

**Comment by paulbutcher** [comment]
I've removed the reference to `tpeDeprecated` and marked this as ready for review. It's definitely a step forwards, and we can enhance as and when we implement #2163.

Should I create a "reformat documentation comments" issue (that will be a fun task for someone 😳)

File: main/src/ca/uwaterloo/flix/language/ast/Symbol.scala [hunk 1]
```diff
@@ -164,6 +172,15 @@ object Symbol {
     new TypeAliasSym(ns.parts, ident.name, ident.loc)
   }
 
+  /**
+    * Returns the type alias symbol for the given fully qualified name
+    */
+
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Symbol.scala:178]: Spurious newline.

File: main/src/ca/uwaterloo/flix/runtime/shell/Command.scala [hunk 2]
```diff
@@ -35,6 +35,11 @@ object Command {
     */
   case object Reload extends Command
 
+  /**
+    * Displays documentation
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/runtime/shell/Command.scala:39]: Prints documentation about the definition or signature with fqn `s`.

File: main/src/ca/uwaterloo/flix/runtime/shell/Shell.scala [hunk 3]
```diff
@@ -172,17 +180,57 @@ class Shell(sourceProvider: SourceProvider, options: Options) {
...
+    * Displays documentation for the given identifier
+    */
+  private def execDoc(s: String)(implicit terminal: Terminal): Unit = {
+    val w = terminal.writer()
+    val classSym = Symbol.mkClassSym(s)
+    val defnSym = Symbol.mkDefnSym(s)
+    val enumSym = Symbol.mkEnumSym(s)
+    val aliasSym = Symbol.mkTypeAliasSym(s)
+
+    for(r <- root) {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/runtime/shell/Shell.scala:205]: I would prefer if this was an explicit match on whether root is None or Some :) (for clarity)

File: main/src/ca/uwaterloo/flix/runtime/shell/Shell.scala [hunk 4]
```diff
@@ -172,17 +180,57 @@ class Shell(sourceProvider: SourceProvider, options: Options) {
...
+        w.println(FormatDoc.asMarkDown(enumDecl.doc))
+
+      } else if (r.typeAliases.contains(aliasSym)) {
+
+        val aliasDecl = r.typeAliases(aliasSym)
+        w.println(FormatType.formatWellKindedType(aliasDecl.tpe))
+        w.println
+        w.println(FormatDoc.asMarkDown(aliasDecl.doc))
+
+      } else {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/runtime/shell/Shell.scala:230]: Remove the extra newlines above.

File: main/src/ca/uwaterloo/flix/runtime/shell/Shell.scala [hunk 5]
```diff
@@ -198,11 +246,12 @@ class Shell(sourceProvider: SourceProvider, options: Options) {
...
     val w = terminal.writer()
 
-    w.println("  Command       Purpose")
+    w.println("  Command       Arguments     Purpose")
     w.println()
-    w.println("  :reload :r    Recompiles every source file.")
-    w.println("  :quit :q      Terminates the Flix shell.")
-    w.println("  :help :h :?   Shows this helpful information.")
+    w.println("  :reload :r                  Recompiles every source file.")
+    w.println("  :doc :d       <identifier>  Displays documentation for <identifier>")
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/runtime/shell/Shell.scala:252]: Maybe we should say `<fqn>` instead of `identifier` ?

**Review by magnus-madsen — COMMENTED** [review]
Overall looks good. I just have some nitpicks.

File: main/src/ca/uwaterloo/flix/runtime/shell/Command.scala (hunk 2 above)
Comment (paulbutcher) [reply main/src/ca/uwaterloo/flix/runtime/shell/Command.scala:39]: Done, but it does more than just definitions and signatures (it also handles enums and aliases).

**Comment by magnus-madsen** [comment]
Dang, forgot the `feat:` part :/

**Comment by magnus-madsen** [comment]
Oh well

**Comment by magnus-madsen** [comment]
> > Oh well
> 
> Should we make a bot to nag people about this? :D

Haha. The return of the bots. Sure why not?

**Pull request merged**